symengine/symengine_rcp.h

```diff
@@ -198,6 +198,73 @@ using Teuchos::print_stack_on_segfault;
 #endif
 
 
+template <class T>
+class EnableRCPFromThis {
+// Public interface
+public:
+    //! Get RCP<T> pointer to self (it will cast the pointer to T)
+    inline RCP<T> rcp_from_this() {
+#if defined(WITH_SYMENGINE_RCP)
+        return rcp(static_cast<T*>(this));
+#else
+        return rcp_static_cast<T>(weak_self_ptr_.create_strong());
+#endif
+    }
+
+    //! Get RCP<const T> pointer to self (it will cast the pointer to const T)
+    inline RCP<const T> rcp_from_this() const {
```

Comment:
@bartlettroscoe I see, thanks a lot for the explanation, it makes a lot more sense to me now. Yes, I agree. So I think the current code in this PR is correct, since these are my signatures:
```c++
    inline RCP<T> rcp_from_this();
    inline RCP<const T> rcp_from_this() const;
```
So if the user passes `const T&`, then the compiler will call the `const` version, which will return `RCP<const T>`, and if the user passes `T&`, then the compiler will call the non const version and returning `RCP<T>`, which I think is exactly what we need here.